### PR TITLE
CORS-2852: Build and package Cluster API and providers

### PR DESCRIPTION
**Pull request by vincepri**
This PR sets the groundwork to build and package Cluster API and its controllers as part of the OpenShift Installer platform.

### CODE REVIEW

**Comment by patrickdillon** [comment]
It would be easier to review this if the vendor changes were in a separate commit. In general, we suggest this approach: https://github.com/openshift/installer/blob/master/docs/dev/dependencies.md#go

Also, we require commit message text like: https://github.com/openshift/installer/blob/master/CONTRIBUTING.md#commit-message-format

File: cluster-api/Makefile [hunk 1]
```diff
@@ -0,0 +1,55 @@
+TARGET_OS_ARCH:=$(shell go env GOOS)_$(shell go env GOARCH)
+
+TFSUBDIRS:=	$(foreach DIR,$(shell find providers -maxdepth 1 -mindepth 1 -type d),$(subst providers/,,$(DIR)))
```
Comment (patrickdillon) [inline cluster-api/Makefile:3]: The `TF-` prefix refers to terraform so let's drop it. Perhaps `PROVSUBDIRS` (provider subdirectories)?

Comment (vincepri) [reply cluster-api/Makefile:3]: Done!

**Comment by vincepri** [comment]
> It would be easier to review this if the vendor changes were in a separate commit. In general, we suggest this approach: https://github.com/openshift/installer/blob/master/docs/dev/dependencies.md#go
> 
> Also, we require commit message text like: https://github.com/openshift/installer/blob/master/CONTRIBUTING.md#commit-message-format

Split in multiple commits, thanks!

File: cluster-api/providers/aws/go.mod [hunk 2]
```diff
@@ -0,0 +1,103 @@
+module openshift/installer/cluster-api/providers/aws
+
+go 1.20
+
+require sigs.k8s.io/cluster-api-provider-aws/v2 v2.2.4
+
+replace sigs.k8s.io/cluster-api-provider-aws/v2 => github.com/vincepri/cluster-api-provider-aws/v2 v2.0.0-20231013205918-8347c5355060
```
Comment (patrickdillon) [inline cluster-api/providers/aws/go.mod]: I'm fine with merging these in the current state, but in the future, we would  use forks here, when appropriate, right?

Comment (vincepri) [reply cluster-api/providers/aws/go.mod]: That's correct yes, we're have a couple of PRs left, this can be removed soon

File: hack/build.sh [hunk 3]
```diff
@@ -36,11 +40,19 @@ fi
...
 MODE="${MODE:-release}"
-# build terraform binaries before setting environment variables since it messes up make
-make -C terraform all
 
-# Copy terraform parts to embedded mirror.
-copy_terraform_to_mirror
+# Build terraform binaries before setting environment variables since it messes up make
+if test "${SKIP_TERRAFORM}" != y
+then
+  make -j8 -C terraform all
```
Comment (patrickdillon) [inline hack/build.sh:47]: It's not clear to me that using the `-j` jobs flag is really enabling the jobs to run in parallel, as the manpages say:
```shell
$ make --help | rg "Allow N jobs"
  -j [N], --jobs[=N]          Allow N jobs at once; infinite jobs with no arg.
```

Comment (vincepri) [reply hack/build.sh]: Make by defaults run every job/target in sequential order, this allows Make to run 8 jobs at once, it still respects dependencies though

Comment (vincepri) [reply hack/build.sh]: In other words, it'll run N jobs in parallel, if they can run in parallel, which our Makefile for both terraform and CAPI allow

Comment (patrickdillon) [reply hack/build.sh]: > Allow N jobs at once; infinite jobs with no arg.

Sounds like `Make` will automatically try to parallelize as many jobs as possible?

File: hack/build-cluster-api.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+sync_envtest() {
+  if [ -f "${CLUSTER_API_BIN_DIR}/kube-apiserver" ]; then
+    version=$("${CLUSTER_API_BIN_DIR}/kube-apiserver" --version || echo "Kubernetes v0.0.0")
+    echo "Found envtest binaries with version: ${version}"
+    if [ "${version}" = "Kubernetes v${ENVTEST_K8S_VERSION}" ]; then
+      return
+    fi
+  fi
+
+  bucket="https://storage.googleapis.com/kubebuilder-tools"
```
Comment (patrickdillon) [inline hack/build-cluster-api.sh:38]: For this also (in addition to the providers) I think we may want to move away from either upstream or at least external dependencies (i.e. we would at least want to mirror these dependencies in our build pipeline). But this is fine for now, just checking alignment for the future.

File: hack/build.sh (hunk 3 above)
Comment (vincepri) [reply hack/build.sh]: Yes but it's defaulted to 1, so there is effectively no parallelization

Comment (patrickdillon) [reply hack/build.sh]: The manpages make it sound like the default is ∞

**Comment by vincepri** [comment]
/hold cancel

Failures seems unrelated

**Pull request merged**